fix: support remark flags

mtfurlan opened the pull request:
This would claim that clang supports
-Wsanitize-address
sanitize-address is actually a remark, so now it lists
-Rsanitize-address

I tested the new output for clang 11 against actual clang, running each
flag like "clang $flag -Werror -Wunknown-argument" and documenting if it
gave an error or not, and now it matches.
(except -Wframe-larger-than= because I didn't pass an argument and it
needed one)

mtfurlan committed:
fix: support remark flags

Right now this claims claing supports things like
-Wsanitize-address
sanitize-address is actually a remark, so it should be
-Rsanitize-address

I tested the new output for clang 11 against actual claing, running each
flag like "clang $flag -Werror -Wunknown-argument" and documenting if it
gave an error or not, and now it matches.
(except -Wframe-larger-than= because I didn't pass an argument and it
needed one)

pkolbus commented on parsers/process_clang_git.py:
```diff
@@ -108,7 +108,7 @@ def parse_clang_info(version: str, target_dir: str, input_dir: str) -> None:
 
     shell(
-        ["llvm-tblgen", "-dump-json", "-I", input_dir, f"{input_dir}/Diagnostic.td"],
+        ["llvm-tblgen-11", "-dump-json", "-I", input_dir, f"{input_dir}/Diagnostic.td"],
```
Why did this change? It breaks `make.sh`.

mtfurlan replied:
Sorry didn't mean to commit that one, that was just to make it work on my system outside of docker so I could test easier.

pkolbus commented on parsers/parse-clang-diagnostic-groups.py:
```diff
@@ -558,6 +573,19 @@ def get_messages(self, enabled_by_default: bool) -> list[str]:
+
+        return True
+
+
```
Extra blank line here. Please run `black` (or `make.sh`) to fix up formatting.

pkolbus commented on parsers/parse-clang-diagnostic-groups.py:
```diff
@@ -386,6 +387,20 @@ def get_messages(self, enabled_by_default: bool) -> list[str]:
+
+        :return: True if the group has remark diagnostics, False otherwise.
+        """
+        rem=False
```
Please spell out `remark`, and run `black` or `make.sh` to fix whitespace.

pkolbus commented on parsers/parse-clang-diagnostic-groups.py:
```diff
@@ -386,6 +387,20 @@ def get_messages(self, enabled_by_default: bool) -> list[str]:
+        for child in self.diagnostics:
+            if child.is_remark:
+                rem=True
+
```
[LLVM's ClangDiagnosticsEmitter](https://github.com/llvm/llvm-project/blob/llvmorg-13.0.0/clang/utils/TableGen/ClangDiagnosticsEmitter.cpp) looks recursively at subgroups; this function needs to do the same (`self.children`).

Also, ClangDiagnosticsEmitter raises an error if a group has both remark and non-remark messages. It would be helpful to raise an exception if this assumption is violated, as it means we need to update this parser.

pkolbus reviewed (changes requested):
Hi @mtfurlan, nice catch, and thanks for this contribution.

A few other things to keep everything clean and consistent:
- The `ClangWarningSwitch` class should probably be renamed (for example to `ClangSwitch` or `ClangDiagnosticSwitch`) and its docstring updated, to reflect that it covers remarks too.
- In the commit message, `clang` is mis-spelled twice. It would also be helpful to distinguish prior behavior with past tense and new behavior as present tense. However, no need to describe how you tested in the commit message.
- Documentation -- the introductory text in `README.md` should be cleaned up to reflect that remark switches are included now. Likewise the text for each compiler's `README.md` (generated from `README_TEMPLATE` in `process_clang_git.py`).
- The xcode warning lists are also affected by this issue and need an update. `process_xcode_git.py` uses the same core parser as clang, so this is a matter of updating the `README_TEMPLATE` and running  `make.sh -x` or equivalent.

As an alternative, maybe remark flags should be omitted from the lists? (This is just a thought -- I've gone back and forth on this point, and would accept either approach).

mtfurlan committed:
code cleanup

ClangWarningSwitch -> ClangDiagnosticSwitch
group is_remark now recursive and throws error if disagreement
linting
remove bad call to llvm-tblgen-11

mtfurlan commented:
For the readme changes I'm finding that I want to replace most instances of "warning" with "diagnostic".
Does that sit right with you?

On one hand I think it's useful to have both, on the other this repo is named "compiler-warnings" not "compiler-diagnostics".

I want this repo to be correct so I can use it as a starting point for some scripts to test flags against apple clang to make an actual table of supported diagnostics.
I don't really care about remarks, but if I'm going to all this trouble I may as well make a complete table.
At $work I've been having fun with a makefile that needs to support a few different versions of clang and apple clang and guessing my way through what warning flags exist has been very frustrating.

mtfurlan commented:
@pkolbus For the readme changes are you fine if I change everything from compiler warning to compiler diagnostic?

If not it might be better to just not include remark flags in the repo.

I don't feel strongly either way but I think it would be easier to do both instead of filtering out the remarks.

pkolbus commented:
Apologies for the long delay in responding and thanks for your patience -- between the holidays, a long-distance relocation, and other things this just fell off my radar.
